File: 18_max_path_sum.py
import os, io
import pdb
import Queue

def createRoot():
    f = open('triangle.txt', 'r')

    root = {}
    queue = Queue.Queue()

    for line in f.readlines():

        line = line.split()
        if not root:
            root[line[0]] = {}
            queue.put(root[line[0]])
            continue

        tmp = queue.get()
        for i in range(len(line)):
            if i == 0 or i == len(line)-1:
                tmp[line[i]] = {}
                queue.put(tmp[line[i]])
            else:
                child = {}
                tmp[line[i]] = child
                queue.put(tmp[line[i]])
                tmp = queue.get()
                tmp[line[i]] = child

        #print line
    return root

def depthFirstSearch(node, maxSum, curSum):
    if not node.keys():
        return curSum

    for key in node.keys():
        result = depthFirstSearch(node[key], maxSum, curSum+int(key))
        if result > maxSum:
            maxSum = result

    return maxSum


root = createRoot()
print depthFirstSearch(root,  0, 0)

File: 18_max_path_sum.rb
require 'pry'

def create_root

  f = File.open('pyramid.txt', 'r').readlines

  f.map! { |e| e.sub("\n", '') }

  root = {}
  # implementing a python FIFO queue using Array#shift and Array#push
  queue = []

  f.each do |line|
    line = line.split

    if root.empty?
      root[line[0]] = {}
      queue.push(root[line[0]])
      next
    end

    tmp = queue.shift

    (0..line.length).each do |i|
      if i == 0 || i == (line.length - 1)
        tmp[line[i]] = {}
        queue.push(tmp[line[i]])
      else
        child = {}
        tmp[line[i]] = child
        queue.push(tmp[line[i]])
        tmp = queue.shift
        tmp[line[i]] = child
      end
    end
  end
  return root
end

def depth_first_search(node, max_sum, cur_sum

end
